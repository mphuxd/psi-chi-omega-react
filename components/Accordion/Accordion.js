import React from "react";
import cx from "classnames";

const Accordion = ({ children, className }) => {
  let classNames = cx(className);
  return <ol className={classNames}>{children}</ol>;
};

export default Accordion;
