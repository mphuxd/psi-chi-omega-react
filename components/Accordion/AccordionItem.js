import React, { useState } from "react";
import cx from "classnames";
import useBreakpointSize from "../../hooks/useBreakpointSize";
import Image from "next/image";

const AccordionItem = ({ title, caption, children, open = false }) => {
  const [isOpen, setIsOpen] = useState(open);
  let className = cx({
    ["accordion__item text--body col-span-full lg:col-span-8 lg:block"]: true,
    ["accordion__item--active block pb-4"]: isOpen,
    ["accordion__item--closed hidden"]: !isOpen,
    ["accordion__item--disabled"]: false,
  });

  let captionClass = cx({
    ["text-left lg:inline"]: true,
    ["inline"]: isOpen,
    ["hidden"]: !isOpen,
  });

  let breakPoint = useBreakpointSize();
  let disabledBreakPoints = ["lg", "xl", "2xl", "max"];
  const onClick = () => {
    if (disabledBreakPoints.includes(breakPoint)) {
      return false;
    }
    setIsOpen(!isOpen);
  };

  return (
    <li className='theme-grid__inner relative py-4 md:py-6'>
      <button
        onClick={onClick}
        className='accordion--item col-span-full lg:col-span-4 flex justify-between flex-row lg:hover:cursor-default'
      >
        <div className='flex flex-col'>
          <h3 className='inline-block font-body text-2xl leading-8 font-bold sm:text-4xl md:text-3xl md:leading-tight xl:leading-tight'>
            {title}
          </h3>
          <p className={captionClass}>{caption}</p>
        </div>
        <div className='inline-block lg:hidden'>
          <Image src='/images/icons/expand_more.svg' alt='Expand More' width={32} height={32} />
        </div>
      </button>

      <div className={className}>{children}</div>
    </li>
  );
};

export default AccordionItem;
