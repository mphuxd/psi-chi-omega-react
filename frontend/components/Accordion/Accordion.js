import cx from "classnames";

export const ACCORDION_TEST_ID = "accordion";

function Accordion({ children, className }) {
  let classNames = cx(className);
  return (
    <ol data-testid={ACCORDION_TEST_ID} className={classNames}>
      {children}
    </ol>
  );
}

export default Accordion;
