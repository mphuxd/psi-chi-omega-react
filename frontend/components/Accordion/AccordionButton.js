import React from "react";
import Image from "next/image";
import cx from "classnames";

export const ACCORDION_BUTTON_TEST_ID = "accordion-button";

function AccordionButton({ isActive, disabledBreakpoints }) {
  let classNames = cx({
    ["accordion-button"]: true,
    [disabledBreakpoints[0] + ":" + "hidden"]: true,
    ["rotate-180"]: isActive,
  });
  return (
    <div
      aria-hidden={true}
      className={classNames}
      data-state={isActive}
      data-testid={ACCORDION_BUTTON_TEST_ID}
    >
      <Image src='/images/icons/expand_more.svg' alt='Expand More' width={32} height={32} />
    </div>
  );
}

export default AccordionButton;
