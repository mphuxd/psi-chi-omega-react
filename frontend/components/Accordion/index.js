import Accordion from './Accordion.js';
import AccordionItem from './AccordionItem.js';

export { Accordion, AccordionItem };
//export both as named exports
