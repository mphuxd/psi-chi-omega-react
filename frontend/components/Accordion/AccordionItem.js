import React, { useState, useRef } from "react";
import AccordionButton from "./AccordionButton";
import useBreakpointSize from "../../hooks/useBreakpointSize";
import cx from "classnames";

export const ACCORDION_ITEM_TEST_ID = "accordion-item";
export const ACCORDION_TRIGGER_TEST_ID = "accordion-trigger";
export const ACCORDION_HEADER_TEST_ID = "accordion-header";
export const ACCORDION_CAPTION_TEST_ID = "accordion-caption";
export const ACCORDION_CONTENT_TEXT_TEST_ID = "accordion-content-text";

const AccordionItem = ({ title, caption, children, startOpened = false, index }) => {
  //atypical accordion item.
  //at these breakpoints, the layout changes & accordion is not collapsible
  const disabledBreakpoints = ["lg", "xl", "2xl", "max"];
  const [isOpen, setIsOpen] = useState(startOpened);
  let isDisabled = useRef(null);

  let accordionItemContentClass = cx({
    ["accordion-item_content"]: true,
    ["accordion-item_content--open"]: isOpen,
    ["accordion-item_content--closed"]: !isOpen,
  });

  let accordionCaptionClass = cx({
    ["accordion-item_caption"]: true,
    ["accordion-item_caption--open"]: isOpen,
    ["accordion-item_caption--closed"]: !isOpen,
  });

  let breakpoint = useBreakpointSize();

  if (disabledBreakpoints.includes(breakpoint)) {
    isDisabled = true;
  } else {
    isDisabled = false;
  }

  if (disabledBreakpoints.includes(breakpoint) && !isOpen) {
    setIsOpen(true);
  }

  const onClick = (e) => {
    if (disabledBreakpoints.includes(breakpoint)) {
      return false;
    }
    setIsOpen(!isOpen);
  };
  
  const accordionID = "accordion-" + index + "-id";
  const ariaControls = title.toLowerCase().replaceAll(" ", "-").concat("-content");

  return (
    <li
      role='heading'
      aria-level={3}
      className='accordion-item'
      data-state={isOpen}
      data-testid={ACCORDION_ITEM_TEST_ID}
    >
      <button
        className='accordion-item_trigger'
        id={accordionID}
        onClick={onClick}
        aria-controls={ariaControls}
        aria-expanded={isOpen}
        aria-disabled={isDisabled}
        data-state={isOpen}
        data-testid={ACCORDION_TRIGGER_TEST_ID}
      >
        <div className='flex flex-col'>
          <span className='accordion-item_header' data-testid={ACCORDION_HEADER_TEST_ID}>
            {title}
          </span>
          {caption && (
            <span className={accordionCaptionClass} data-testid={ACCORDION_CAPTION_TEST_ID}>
              {caption}
            </span>
          )}
        </div>
        <AccordionButton isActive={isOpen} disabledBreakpoints={disabledBreakpoints} />
      </button>

      <div
        id={ariaControls}
        aria-labelledby={accordionID}
        className={accordionItemContentClass}
        data-testid={ACCORDION_CONTENT_TEXT_TEST_ID}
      >
        {children}
      </div>
    </li>
  );
};

export default AccordionItem;
