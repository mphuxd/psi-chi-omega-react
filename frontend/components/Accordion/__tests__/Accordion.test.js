import { render, screen } from "@testing-library/react";
import Accordion, { ACCORDION_TEST_ID } from "../Accordion";
import AccordionItem from "../AccordionItem";

describe("Accordion ", () => {
  it("should render custom classnames", async () => {
    const className = "class-test";
    render(
      <Accordion className={className}>
        <AccordionItem title='Section 1 title' caption='caption-1'>
          content
        </AccordionItem>
        <AccordionItem title='Section 2 title' caption='caption-2'>
          content 2
        </AccordionItem>
        <AccordionItem title='Section 3 title' caption='caption-3'>
          content 3
        </AccordionItem>
      </Accordion>
    );

    const accordion = await screen.findByTestId(ACCORDION_TEST_ID);
    expect(accordion).toHaveClass(className);
    expect(accordion).toMatchSnapshot();
  });
});
