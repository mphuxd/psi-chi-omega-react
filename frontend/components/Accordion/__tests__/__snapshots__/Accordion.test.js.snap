// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion  should render custom classnames 1`] = `
<ol
  class="class-test"
  data-testid="accordion"
>
  <li
    class="accordion-item"
    data-state="true"
    data-testid="accordion-item"
  >
    <button
      class="accordion-item_trigger"
      data-state="true"
      data-testid="accordion-trigger"
    >
      <div
        class="flex flex-col"
      >
        <h3
          class="accordion-item_header"
          data-testid="accordion-header"
        >
          Section 1 title
        </h3>
        <p
          class="accordion-item_caption accordion-item_caption--open"
          data-testid="accordion-caption"
        >
          caption-1
        </p>
      </div>
      <div
        class="accordion-button lg:hidden rotate-180"
        data-state="true"
        data-testid="accordion-button"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2732%27%20height=%2732%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            alt="Expand More"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
    </button>
    <div
      class="accordion-item_content accordion-item_content--open"
      data-testid="accordion-content-text"
    >
      content
    </div>
  </li>
  <li
    class="accordion-item"
    data-state="true"
    data-testid="accordion-item"
  >
    <button
      class="accordion-item_trigger"
      data-state="true"
      data-testid="accordion-trigger"
    >
      <div
        class="flex flex-col"
      >
        <h3
          class="accordion-item_header"
          data-testid="accordion-header"
        >
          Section 2 title
        </h3>
        <p
          class="accordion-item_caption accordion-item_caption--open"
          data-testid="accordion-caption"
        >
          caption-2
        </p>
      </div>
      <div
        class="accordion-button lg:hidden rotate-180"
        data-state="true"
        data-testid="accordion-button"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2732%27%20height=%2732%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            alt="Expand More"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
    </button>
    <div
      class="accordion-item_content accordion-item_content--open"
      data-testid="accordion-content-text"
    >
      content 2
    </div>
  </li>
  <li
    class="accordion-item"
    data-state="true"
    data-testid="accordion-item"
  >
    <button
      class="accordion-item_trigger"
      data-state="true"
      data-testid="accordion-trigger"
    >
      <div
        class="flex flex-col"
      >
        <h3
          class="accordion-item_header"
          data-testid="accordion-header"
        >
          Section 3 title
        </h3>
        <p
          class="accordion-item_caption accordion-item_caption--open"
          data-testid="accordion-caption"
        >
          caption-3
        </p>
      </div>
      <div
        class="accordion-button lg:hidden rotate-180"
        data-state="true"
        data-testid="accordion-button"
      >
        <span
          style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
        >
          <span
            style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
          >
            <img
              alt=""
              aria-hidden="true"
              src="data:image/svg+xml,%3csvg%20xmlns=%27http://www.w3.org/2000/svg%27%20version=%271.1%27%20width=%2732%27%20height=%2732%27/%3e"
              style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
            />
          </span>
          <img
            alt="Expand More"
            data-nimg="intrinsic"
            decoding="async"
            src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
            style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
          />
          <noscript />
        </span>
      </div>
    </button>
    <div
      class="accordion-item_content accordion-item_content--open"
      data-testid="accordion-content-text"
    >
      content 3
    </div>
  </li>
</ol>
`;
